clang/lib/CConv/DeclRewriter.cpp

```diff
@@ -0,0 +1,410 @@
+                                                  NewFuncSig, ABRewriter);
+  StructVariableInitializer SVI = StructVariableInitializer(&Context, Info,
+                                                            RewriteThese);
+  for (const auto &D : Context.getTranslationUnitDecl()->decls()) {
+    TRV.TraverseDecl(D);
+    SVI.TraverseDecl(D);
+  }
+
+  // Build a map of all of the PersistentSourceLoc's back to some kind of
+  // Stmt, Decl, or Type.
+  TranslationUnitDecl *TUD = Context.getTranslationUnitDecl();
+  std::set<PersistentSourceLoc> Keys;
+  for (const auto &I : Info.getVarMap())
+    Keys.insert(I.first);
+  MappingVisitor MV(Keys, Context);
+  for (const auto &D : TUD->decls())
+    MV.TraverseDecl(D);
+  SourceToDeclMapType PSLMap;
+  VariableDecltoStmtMap VDLToStmtMap;
+  std::tie(PSLMap, VDLToStmtMap) = MV.getResults();
+
+  // Add declarations from this map into the rewriting set
+  for (const auto &V : Info.getVarMap()) {
+    PersistentSourceLoc PLoc = V.first;
+    CVarSet Vars = V.second;
+    // I don't think it's important that Vars have any especial size, but
+    // at one point I did so I'm keeping this comment here. It's possible
+    // that what we really need to do is to ensure that when we work with
+    // either PV or FV below, that they are the LUB of what is in Vars.
+    // assert(Vars.size() > 0 && Vars.size() <= 2);
+
+    // PLoc specifies the location of the variable whose type it is to
+    // re-write, but not where the actual type storage is. To get that, we
+    // need to turn PLoc into a Decl and then get the SourceRange for the
+    // type of the Decl. Note that what we need to get is the ExpansionLoc
+    // of the type specifier, since we want where the text is printed before
+    // the variable name, not the typedef or #define that creates the
+    // name of the type.
+    if (Decl *D = std::get<1>(PSLMap[PLoc])) {
+      // We might have one Decl for multiple Vars, however, one will be a
+      // PointerVar so we'll use that.
+      PVConstraint *PV = nullptr;
+      FVConstraint *FV = nullptr;
+      for (const auto &V : Vars) {
+        if (PVConstraint *T = dyn_cast<PVConstraint>(V))
+          PV = T;
+        else if (FVConstraint *T = dyn_cast<FVConstraint>(V))
+          FV = T;
+      }
+
+      if (PV && PV->anyChanges(Info.getConstraints().getVariables()) &&
+          !PV->isPartOfFunctionPrototype()) {
+        // Rewrite a declaration, only if it is not part of function prototype.
+        DeclStmt *DS = nullptr;
+        if (VDLToStmtMap.find(D) != VDLToStmtMap.end())
+          DS = VDLToStmtMap[D];
+
+        std::string newTy = getStorageQualifierString(D) +
+            PV->mkString(Info.getConstraints().getVariables()) +
+            ABRewriter.getBoundsString(PV, D);
+        RewriteThese.insert(DAndReplace(D, DS, newTy));
+      } else if (FV && NewFuncSig.find(FV->getName()) != NewFuncSig.end()
+                    && !TRV.isFunctionVisited(FV->getName())) {
```

Comment:
@mwhicks1 I've realized that this branch of the conditional is never taken in our test or when converting vsftpd. Removing it would simplify some things a decent amount, but I'm not convinced it will never be taken. Thoughts?

Reply:
Per our call yesterday, we will just keep this, right? I think there are times when not-so-great C code will use a function before it's declared, and `clang` will let it compile.

Reply:
Yes, I plan to keep this branch, but I want to find a test case that actually hits it.